lib/request-proxy-send.js

```diff
@@ -0,0 +1,279 @@
+    this.numRetries++;
+
+    // TODO Support this case too
+    var dests = this.lookupKeys(this.keys);
+    if (dests.length > 1) {
+        this.abortRetry(dests, callback);
+        return;
+    }
+
+    this.ringpop.stat('increment', 'requestProxy.retry.attempted');
+    this.ringpop.emit('requestProxy.retryAttempted');
+
+    var newDest = dests[0];
+
+    // Nothing rebalanced.
+    if (newDest === this.channelOpts.host) {
+        this.send(this.channelOpts, callback);
+        return;
+    }
+
+    this.rerouteRetry(newDest, callback);
+};
+
+RequestProxySend.prototype.destroy = function destroy() {
+    clearTimeout(this.timeout);
+};
+
+RequestProxySend.prototype.getRawHead = function getRawHead() {
+    return rawHead(this.request.obj, {
+        checksum: this.ringpop.membership.checksum,
+        keys: this.keys
+    });
+};
+
+RequestProxySend.prototype.getStrHead = function getStrHead() {
+    return strHead(this.request.obj, {
+        checksum: this.ringpop.membership.checksum,
+        keys: this.keys
+    });
+};
+
+RequestProxySend.prototype.handleMaxRetriesExceeded = function handleMaxRetriesExceeded(callback) {
+    this.ringpop.stat('increment', 'requestProxy.retry.failed');
+    this.ringpop.logger.warn('ringpop request proxy retry exceeded max retries and failed', {
+        local: this.ringpop.whoami(),
+        errors: this.errors,
+        maxRetries: this.maxRetries,
+        maxRetryTimeout: this.maxRetryTimeout,
+        reqMethod: this.reqMethod,
+        reqUrl: this.reqUrl,
+        reqStartTime: this.reqStartTime,
+        reqTotalTime: Date.now() - this.reqStartTime,
+        retrySchedule: this.retrySchedule,
+        retryStartTime: this.retryStartTime,
+        retryTotalTime: Date.now() - this.retryStartTime,
+        destinations: this.destinations
+    });
+
+    this.ringpop.emit('requestProxy.retryFailed');
+
+    callback(MaxRetriesExceeded({
+        maxRetries: this.maxRetries
+    }));
+};
+
+RequestProxySend.prototype.handleSuccess = function handleSuccess(res1, res2, callback) {
+    if (this.numRetries > 0) {
+        this.ringpop.stat('increment', 'requestProxy.retry.succeeded');
+        this.ringpop.logger.info('ringpop request proxy retry succeeded', {
+            local: this.ringpop.whoami(),
+            errors: this.errors,
+            numRetries: this.numRetries,
+            maxRetries: this.maxRetries,
+            reqMethod: this.reqMethod,
+            reqUrl: this.reqUrl,
+            reqStartTime: this.reqStartTime,
+            reqTotalTime: Date.now() - this.reqStartTime,
+            retrySchedule: this.retrySchedule,
+            retryStartTime: this.retryStartTime,
+            retryTotalTime: Date.now() - this.retryStartTime,
+            destinations: this.destinations
+        });
+
+        this.ringpop.emit('requestProxy.retrySucceeded');
+    }
+
+    callback(null, res1, res2);
+};
+
+RequestProxySend.prototype.lookupKeys = function lookupKeys(keys) {
```

Comment:
nitpick: doing a reduce and Object.keys() feels a bit strange.

You probably wanted

```js
var dests = [];

for (var i = 0; i < keys.length; i++) {
    var key = self.ringpop.lookup(keys[i]);
    if (dests.indexOf(key) === -1) {
        dests.push(key);
    }
}

return dests;
```